skimage/restoration/_nl_means_denoising.pyx

```diff
@@ -0,0 +1,596 @@
+import numpy as np
+from skimage import util
+cimport numpy as np
+cimport cython
+from libc.math cimport exp
+
+ctypedef np.float32_t DTYPE_t
+
+cdef eps = 1.e-8
+
+
+@cython.boundscheck(False)
+cdef inline float patch_distance_2d(DTYPE_t [:, :] p1,
+                                    DTYPE_t [:, :] p2,
+                                    DTYPE_t [:, ::] w, int s):
```

Comment:
These three patch distance functions can be replaced by a single nD function:

```python
@cython.boundscheck(False)
cdef inline float patch_distance_r(DTYPE_t [::1] p1,
                                   DTYPE_t [::1] p2,
                                   DTYPE_t [::1] w, int s):
    cdef int i
    cdef int center = s / 2
    cdef float tmp_diff = p1[center] - p2[center]
    cdef float init = w[center] * tmp_diff * tmp_diff
    if init > 1:
        return 0.
    cdef float distance = 0
    for i in range(s):
        if distance > 5:
            return 0.
        tmp_diff = p1[i] - p2[i]
        distance += w[i] * tmp_diff * tmp_diff
    return exp(-distance)
```

Then instead of passing in `patch[i0:i1, j0:j1]`, you pass `patch[i0:i1, j0:j1].ravel()`, and make sure s is the square of the current definition, ie for patch size 7 you pass in 49, etc.

I thought that this might adversely affect performance, but it seems to have almost no effect. Before:

```python
In [5]: %timeit denoise = nl_means_denoising(noisy, 7, 9, 0.08)
1 loops, best of 3: 3.25 s per loop
```

after:

```python
In [5]: %timeit denoise = nl_means_denoising(noisy, 7, 9, 0.08)
1 loops, best of 3: 3.27 s per loop
```